docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+| Attributes | map | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.|
+| *Device Command Properties* ||||
+| Name | string | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  It can never be blank and must be unique across the entire profile. |
+| IsHidden | bool | Anytime | This field allows the device command to be exposed (default) via the command service.  This change does not affect the behavior but only visibility.  This field is optional but is false by default when not provided.|
+| ReadWrite | string | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  This field is required.  As the field is expected to be from an enumeration of R, W and RW, it can never be empty. |
+| ResourceOperations | array | Situational | When the device profile is not associated to any device, reading or event, this field can be modified.  This field is required and can never be empty. |
+
+ 
+### Empty Profile
+
+A device profile can begin life “empty” - meaning that it has no device resources or device commands.  Once associated to any object (device, event, reading) then, since device resource or device command additions can be made, the empty profile can gain resources or commands.  You can never remove device resources or device commands once the profile is associated to another object.
+
+The general properties of a device profile, with the exception of name, are optional when creating an empty profile.
+
+All profiles – even empty profiles – must have a name.                                                                                 
+
+### Deleting Profiles
+
+A profile can only be deleted from metadata when the profile is not associated to any other EdgeX object to include device, provision watcher, event or reading.
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
```

Comment:
I'm not sure I understand the justification here? If a device's P`rofileName `is changed, the new profile will be used to generate new events and readings. Since the existing events and readings still reference the old profile (which can only change according to the given rules), it still would be possible to retrieve additional information from the profile if needed by an application.

Reply:
Agreed.